contracts/staking/EpochSelector.sol

```diff
@@ -0,0 +1,1683 @@
+        uint256 total;
+        for (uint256 index = 0; index < _nodes.length; index++) {
+            Node memory node = nodes[_nodes[index]];
+            total += node.balance;
+        }
+        return total;
+    }
+
+    // function _printNode(address _node) internal view {
+    //     Node memory node = nodes[_node];
+    //     console2.log("************************************");
+    //     console2.log("cluster", node.node);
+    //     console2.log("balance", node.balance);
+    //     console2.log("left", node.left);
+    //     console2.log("right", node.right);
+    //     console2.log("sumOfLeftBalances", node.sumOfLeftBalances);
+    //     console2.log("sumOfRightBalances", node.sumOfRightBalances);
+    //     console2.log(" height", node.height);
+    //     console2.log("************************************");
+    // }
+
+    // function _printArray(string memory data, bytes memory arrayBytes) internal view {
+    //     console2.log(data);
+    //     address[] memory array = abi.decode(arrayBytes, (address[]));
+    //     console2.log("[");
+    //     for (uint256 index = 0; index < array.length; index++) {
+    //         console2.log(index, array[index]);
+    //     }
+    //     console2.log("]");
+    // }
+
+    // function _printArray(string memory data, address[] memory array) internal view {
+    //     console2.log(data);
+    //     console2.log("[");
+    //     for (uint256 index = 0; index < array.length; index++) {
+    //         console2.log(index, array[index]);
+    //     }
+    //     console2.log("]");
+    // }
+
+    // function _printPaths(string memory data, bytes[] memory bytesdata) internal view {
+    //     console2.log(data);
+
+    //     console2.log("[");
+    //     for (uint256 index = 0; index < bytesdata.length; index++) {
+    //         address[] memory _paths = abi.decode(bytesdata[index], (address[]));
+    //         _printArray("subarray ", _paths);
+    //     }
+    //     console2.log("]");
+    // }
+}
+
+contract SingleSelector is SelectorHelper {
+    constructor() {}
+
+    /// @inheritdoc IClusterSelector
+    function insert(address newNode, uint256 balance) public virtual override {
+        require(newNode != address(0), ClusterLib.CANNOT_BE_ADDRESS_ZERO);
+        Node memory node = nodes[newNode];
+        if (node.node == address(0)) {
+            root = _insert(root, newNode, balance);
+            totalElements++;
+        } else {
+            // int256 differenceInKeyBalance = int256(clusterBalance) - int256(node.balance);
+            _update(root, newNode, int256(balance) - int256(node.balance));
+        }
+    }
+
+    /// @inheritdoc IClusterSelector
+    function deleteNode(address key) public virtual override {
+        require(key != address(0), ClusterLib.CANNOT_BE_ADDRESS_ZERO);
+        Node memory node = nodes[key];
+        if (node.node == key) {
```

Comment:
Why not `require` instead of `if` as deletion of a  non existing node is a error right ?